### ISSUE TITLE
Uncaught TypeError: Cannot read property 'unmodified' of undefined

### ISSUE THREAD

**Issue by ZaFaR97** [atom/atom#13296]
**Steps to reproduce**

Prerequisites: Ubuntu 16.04 with Gnome.

1. Open `Region & Lanugage` settings from the Gnome settings.
1. Add the `Thai (TIS-820.2538)` layout and make sure it is the top layout in the `Region & Language` settings.
1. Open Atom and type any character.

This issue was previously fixed for Windows in https://github.com/atom/keyboard-layout/pull/23 but seems to also be an issue on Linux.

![screenshot from 2016-12-25 21-33-32](https://cloud.githubusercontent.com/assets/1058982/21472801/314d5804-caea-11e6-823c-b30571d98e5d.png)

----

IF i press any shortcut such as CTRL + C or CTRL + P
i see this msg : 
Uncaught TypeError: Cannot read property 'unmodified' of undefined

**Atom Version**: 1.12.5
**Electron Version**: 1.3.9
**System**: Microsoft Windows 7 Home Premium 
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: Cannot read property 'unmodified' of undefined

```
At C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\node_modules\atom-keymap\lib\helpers.js:45

TypeError: Cannot read property 'unmodified' of undefined
    at isLatinKeymap (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\node_modules\atom-keymap\lib\helpers.js:45:45)
    at exports.keystrokeForKeyboardEvent (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\node_modules\atom-keymap\lib\helpers.js:243:31)
    at KeymapManager.module.exports.KeymapManager.keystrokeForKeyboardEvent (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\node_modules\atom-keymap\lib\keymap-manager.js:433:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\node_modules\atom-keymap\lib\keymap-manager.js:321:24)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\src\window-event-handler.js:106:36)
    at HTMLDocument.<anonymous> (C:\Users\Z\AppData\Local\atom\app-1.12.5\resources\app.asar\src\window-event-handler.js:3:59)
```

### Commands

```
     -8:52.1.0 command-palette:toggle (atom-text-editor.editor.is-focused)
 10x -8:46.1.0 core:backspace (atom-text-editor.editor.mini.is-focused)
     -8:35 application:open-your-keymap (atom-text-editor.editor)
 43x -8:32.3.0 core:move-down (atom-text-editor.editor.is-focused)
undefined -8:30.3.0 core:move-up (atom-text-editor.editor.is-focused)
```

### Config

```json
{}
```

### Installed Packages

```coffee
# User
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.52.0 (inactive)
atom-light-syntax, v0.28.0 (inactive)
atom-light-ui, v0.45.0 (inactive)
base16-tomorrow-dark-theme, v1.3.0 (inactive)
base16-tomorrow-light-theme, v1.3.0 (inactive)
one-dark-ui, v1.6.2 (active)
one-light-ui, v1.6.2 (inactive)
one-dark-syntax, v1.5.0 (active)
one-light-syntax, v1.5.0 (inactive)
solarized-dark-syntax, v1.0.5 (inactive)
solarized-light-syntax, v1.0.5 (inactive)
about, v1.7.0 (active)
archive-view, v0.62.0 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.13.1 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.4 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.1 (active)
bookmarks, v0.42.0 (active)
bracket-matcher, v0.82.2 (active)
command-palette, v0.39.0 (active)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
exception-reporting, v0.40.0 (active)
find-and-replace, v0.202.2 (active)
fuzzy-finder, v1.4.0 (active)
git-diff, v1.1.0 (active)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.2 (active)
image-view, v0.60.0 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.2 (inactive)
markdown-preview, v0.158.8 (active)
metrics, v1.1.2 (active)
notifications, v0.65.1 (active)
open-on-github, v1.2.1 (inactive)
package-generator, v1.0.1 (inactive)
settings-view, v0.243.1 (active)
snippets, v1.0.4 (active)
spell-check, v0.68.4 (active)
status-bar, v1.4.1 (active)
styleguide, v0.47.2 (active)
symbols-view, v0.113.1 (inactive)
tabs, v0.103.0 (active)
timecop, v0.33.2 (active)
tree-view, v0.210.0 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.35.1 (active)
whitespace, v0.35.0 (active)
wrap-guide, v0.38.2 (active)
language-c, v0.54.0 (active)
language-clojure, v0.22.1 (active)
language-coffee-script, v0.48.0 (active)
language-csharp, v0.13.0 (active)
language-css, v0.40.1 (active)
language-gfm, v0.88.0 (active)
language-git, v0.15.0 (active)
language-go, v0.43.0 (active)
language-html, v0.47.1 (active)
language-hyperlink, v0.16.1 (active)
language-java, v0.24.0 (active)
language-javascript, v0.122.0 (active)
language-json, v0.18.3 (active)
language-less, v0.29.6 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.37.0 (active)
language-php, v0.37.3 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.1 (active)
language-ruby, v0.70.2 (active)
language-ruby-on-rails, v0.25.1 (active)
language-sass, v0.57.0 (active)
language-shellscript, v0.23.0 (active)
language-source, v0.9.0 (active)
language-sql, v0.25.0 (active)
language-text, v0.7.1 (active)
language-todo, v0.29.1 (active)
language-toml, v0.18.1 (active)
language-xml, v0.34.12 (active)
language-yaml, v0.27.1 (active)

# Dev
No dev packages
```

**Comment by as-cii** [comment]
Hey @ZaFaR97, thanks so much for the feedback! It would be great if you could provide more information to help us understand and fix the issue you are experiencing:

* What's the language of your keyboard?
* When did you start experiencing this issue? (e.g. prior to Atom 1.12.5, after upgrading to 1.12.0, etc.)
* Did the problem happen after switching keyboard language while Atom was open?
* Can you reproduce when running Atom in safe mode? (`atom --safe` in a command prompt)

Thank you in advance for your time and help!

**Comment by ZaFaR97** [comment]
> What's the language of your keyboard?

AR & EN
> When did you start experiencing this issue? (e.g. prior to Atom 1.12.5, after upgrading to 1.12.0, etc.)

Suddenly, the problem appeared today
> Did the problem happen after switching keyboard language while Atom was open?

At Writing Any character except space and arrows ! " in arabic And En languages "
> Can you reproduce when running Atom in safe mode? (atom --safe in a command prompt)

I deleted all plugins / themes & run atom in safe mode, But same problem occurs
And I test install beta13 Atom, same problem

**Comment by as-cii** [comment]
Unfortunately, I cannot reproduce this locally. Can you post here which Arabic language are you using? I have tried with Arabic (Tunisia) and Arabic (Egypt) but they both seem to work fine.

Also, we have recently tracked down an issue that is potentially related to the problem you're experiencing. To understand if they are the same, can you tell us if it works properly when trying the following steps:

* Quit Atom.
* Switch the language to Arabic in the Windows task bar.
* Open Atom. 

Thanks!

**Comment by ZaFaR97** [comment]
@as-cii 
i Use saudi Arabia Arabic
if Write Any character OR Number in any language i have same error !

i quit Atom and switch to Arabic and Open it,
i Have same problem

**Simply** when I write anything in the Atom Any where error appears
for example : in settings / install packages search space

**Comment by Ben3eeE** [comment]
@as-cii I can reproduce this when installing the head of 1.12 releases but ****not**** on master after running `script\build --install`. Is there some commit missing in 1.12 releases?

Windows 10, x64.

1. Change language to `Arabic (Saudi Arabia)` in the windows language settings, I was able to reproduce with other Arabic languages so it is not limited to Saudi Arabia.
1. Change to `Arabic (Saudi Arabia)` on the Windows task bar
1. Launch atom using `atom --safe`
1. Press any key.

![image](https://cloud.githubusercontent.com/assets/1058982/20541066/ef45381c-b0fb-11e6-89cf-cec2edb4db6e.png)

**Comment by mitosandov** [comment]
I got the same issue - my keyboard is bulgarian + USA english... this issue happens when i save a file, but the file is saved sucessfully...

**Comment by Ben3eeE** [comment]
@mitosandov I am not sure that you are facing the same issue. This problems occurs for every keystroke not for a specific action like saving the file. Can you please share the full details from the exception notification, preferably in a separate issue so we can take a look, thanks.

**Comment by mitosandov** [comment]
@Ben3eeE  i just fixed it - the only change i made yesterday evening to my computer was to install java from oracle's website. I just uninstalled it and the athom now works perfectly fine. For me issue is solved! Thanks

**Comment by ZaFaR97** [comment]
**Thanks** to all ` ;`

**Comment by as-cii** [comment]
A fix for this has landed on master and it will go out with the next hotfixes. Thanks for your great help! 🙇

**Comment by kwstasna** [comment]
Hello. I also have this error but i'm new to atom and i dont know how to run `--update` so i can skip this error? I have the vestion 1.12.5

**Comment by Petya-Naumova** [comment]
I can cofirm that the issue is present to me. 

> Uncaught TypeError: Cannot read property 'unmodified' of undefined
 My Atom's version is 1.12.5 and it says it's the latest version when I try to update it through the editor. The error occurs when I try to save - Ctrl+S a file.

**Comment by Ben3eeE** [comment]
Hey, This issue has been fixed on master and will go out with the next hotfixes which will be Atom 1.12.6. What @ZaFaR97 meant was that he installed Atom 1.12.1 and that version worked for him.

Sorry for the inconvenience.

**Comment by kwstasna** [comment]
Actually in Windows 10, what i did was to open cmd and run `atom --update` and the issue was fixed 😆

**Comment by Shayennn** [comment]
This issue is still present to me when I try to type Thai.

**Atom Version**: 1.12.7
**Electron Version**: 1.3.13
**System**: Ubuntu 16.04.1
**Thrown From**: Atom Core


### Stack Trace

Uncaught TypeError: Cannot read property 'unmodified' of undefined

```
At /usr/share/atom/resources/app.asar/node_modules/atom-keymap/lib/helpers.js:45

TypeError: Cannot read property 'unmodified' of undefined
    at isLatinKeymap (/usr/share/atom/resources/app.asar/node_modules/atom-keymap/lib/helpers.js:45:45)
    at exports.keystrokeForKeyboardEvent (/usr/share/atom/resources/app.asar/node_modules/atom-keymap/lib/helpers.js:234:31)
    at KeymapManager.module.exports.KeymapManager.keystrokeForKeyboardEvent (/usr/share/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:433:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/usr/share/atom/resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:321:24)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/usr/share/atom/resources/app.asar/src/window-event-handler.js:106:36)
    at HTMLDocument.<anonymous> (/usr/share/atom/resources/app.asar/src/window-event-handler.js:3:59)
```

### Commands

```
```

### Config

```json
{
  "core": {
    "autoHideMenuBar": true,
    "disabledPackages": [
      "file-type-icons",
      "activate-power-mode"
    ],
    "telemetryConsent": "limited",
    "themes": [
      "one-dark-ui",
      "base16-tomorrow-dark-theme"
    ]
  },
  "editor": {
    "invisibles": {},
    "showIndentGuide": true,
    "softWrap": true
  }
}
```

### Installed Packages

```coffee
# User
activate-power-mode, v1.1.0 (inactive)
afterglow-monokai-syntax, v1.3.4 (inactive)
atom-beautify, v0.29.14 (inactive)
atom-bootstrap3, v1.2.12 (inactive)
ba-dum-tss, v1.0.0 (inactive)
codeigniter, v0.3.1 (active)
color-picker, v2.2.3 (active)
docblockr, v0.9.1 (active)
emmet, v2.4.3 (active)
file-icons, v1.7.25 (active)
git-control, v0.8.2 (inactive)
git-plus, v6.0.1 (active)
git-time-machine, v1.5.4 (inactive)
highlight-selected, v0.11.2 (active)
linter, v1.11.18 (active)
linter-php, v1.3.1 (active)
minimap, v4.25.7 (active)
trailing-spaces, v0.3.2 (active)
tree-view-git-status, v1.3.0 (active)
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.52.0 (inactive)
atom-light-syntax, v0.28.0 (inactive)
atom-light-ui, v0.45.0 (inactive)
base16-tomorrow-dark-theme, v1.3.0 (active)
base16-tomorrow-light-theme, v1.3.0 (inactive)
one-dark-ui, v1.6.2 (active)
one-light-ui, v1.6.2 (inactive)
one-dark-syntax, v1.5.0 (inactive)
one-light-syntax, v1.5.0 (inactive)
solarized-dark-syntax, v1.0.5 (inactive)
solarized-light-syntax, v1.0.5 (inactive)
about, v1.7.0 (active)
archive-view, v0.62.0 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.13.1 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.4 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.1 (active)
bookmarks, v0.42.0 (active)
bracket-matcher, v0.82.2 (active)
command-palette, v0.39.0 (inactive)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
exception-reporting, v0.40.0 (active)
find-and-replace, v0.202.2 (inactive)
fuzzy-finder, v1.4.0 (active)
git-diff, v1.1.0 (active)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.2 (active)
image-view, v0.60.0 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.2 (inactive)
markdown-preview, v0.158.8 (active)
metrics, v1.1.2 (active)
notifications, v0.65.1 (active)
open-on-github, v1.2.1 (inactive)
package-generator, v1.0.1 (inactive)
settings-view, v0.243.1 (active)
snippets, v1.0.4 (active)
spell-check, v0.68.4 (active)
status-bar, v1.4.1 (active)
styleguide, v0.47.2 (active)
symbols-view, v0.113.1 (inactive)
tabs, v0.103.0 (active)
timecop, v0.33.2 (active)
tree-view, v0.210.0 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.35.1 (active)
whitespace, v0.35.0 (active)
wrap-guide, v0.38.2 (active)
language-c, v0.54.0 (active)
language-clojure, v0.22.1 (active)
language-coffee-script, v0.48.0 (active)
language-csharp, v0.13.0 (active)
language-css, v0.40.1 (active)
language-gfm, v0.88.0 (active)
language-git, v0.15.0 (active)
language-go, v0.43.0 (active)
language-html, v0.47.1 (active)
language-hyperlink, v0.16.1 (active)
language-java, v0.24.0 (active)
language-javascript, v0.122.0 (active)
language-json, v0.18.3 (active)
language-less, v0.29.6 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.37.0 (active)
language-php, v0.37.3 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.1 (active)
language-ruby, v0.70.2 (active)
language-ruby-on-rails, v0.25.1 (active)
language-sass, v0.57.0 (active)
language-shellscript, v0.23.0 (active)
language-source, v0.9.0 (active)
language-sql, v0.25.0 (active)
language-text, v0.7.1 (active)
language-todo, v0.29.1 (active)
language-toml, v0.18.1 (active)
language-xml, v0.34.12 (active)
language-yaml, v0.27.1 (active)

# Dev
No dev packages
```

**Comment by Ben3eeE** [comment]
Reopening while we investigate this issue still happening on Ubuntu with Thai layout.

@Shayennn 
* Can you reproduce this issue in safe mode `atom --safe`?
* Can you provide the exact name of the Layout you are using
* What is your current Window Manager?

**Comment by Shayennn** [comment]
@Ben3eeE 

Can you reproduce this issue in safe mode atom --safe?

- Yes, It stills present to me in safe mode.

Can you provide the exact name of the Layout you are using?

- Thai (TIS-820.2538)

What is your current Window Manager?

- It's Compiz.

Did you install Atom from https://atom.io/ or via the PPA?

- I download and install .deb file from https://atom.io/.

**Comment by Ben3eeE** [comment]
I can reproduce this on Ubuntu 16.04 Gnome using `Atom 1.14.0-dev-b219c28` with the `Thai (TIS-820.2538)` layout. When this layout is on the top of the input sources this exception happens on every keystroke. This issue was previously fixed for Windows in https://github.com/atom/keyboard-layout/pull/23 but seems to also be an issue on Linux.

![screenshot from 2016-12-25 21-33-32](https://cloud.githubusercontent.com/assets/1058982/21472801/314d5804-caea-11e6-823c-b30571d98e5d.png)

**Comment by Peekungstory** [comment]
atom's developers fix it please.
it still error when type thai language.
*version 1.12.9

**Comment by as-cii** [comment]
This got fixed on master with #13543 and will be available starting with Atom 1.14.0-beta0. Thanks everyone for the reports!

**Comment by itaniknight** [comment]
I am getting this error in **Atom 1.14.0-beta1** in a specific circumstance of my own devising!

### Steps to reproduce

1. Decide you want to be able to use `altgr + e/s/d/f` as a replacement for the arrow keys. (Actually, `./o/e/u` since I'm using Dvorak, but I don't believe that's what's causing the error.)
2. Use xmodmap to bind `altgr` to `Mode_switch` and `e/s/d/f` to the arrow keys in the switched mode.
3. Hold `altgr` (now `Mode_switch`) and press any key *other* than `e/s/d/f`.
4. Receive "cannot read property" error.

My .Xmodmap file looks like this:

```
keycode 108 = Mode_switch

keycode 26 = period greater Up
keycode 39 = o O Left
keycode 40 = e E Down
keycode 41 = u U Right

keycode 53 = q Q Home
keycode 55 = k K End
```

`keycode 39 = o O Left` means: type `o` with no modifiers, `O` when `shift` is held, and `Left` when holding `Mode_switch`.

I'm pretty sure using `Mode_switch` with all the other, *unspecified* key combinations is what's causing the error, since Atom (understandably) doesn't know what `Mode_switch + [???]` means.

### System

**Atom**: 1.14.0-beta1 x64
**Electron**: 1.3.13
**OS**: Ubuntu 16.10
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: Cannot read property 'unmodified' of undefined

```
At /usr/share/atom-beta/resources/app.asar/node_modules/atom-keymap/lib/helpers.js:39

TypeError: Cannot read property 'unmodified' of undefined
    at isLatinKeymap (/app.asar/node_modules/atom-keymap/lib/helpers.js:39:89)
    at exports.keystrokeForKeyboardEvent (/app.asar/node_modules/atom-keymap/lib/helpers.js:228:31)
    at KeymapManager.module.exports.KeymapManager.keystrokeForKeyboardEvent (/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:445:14)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:325:24)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/app.asar/src/window-event-handler.js:106:36)
    at /app.asar/src/window-event-handler.js:3:59)
```

### Non-Core Packages

```
gruvbox 1.0.5 
language-lua 0.9.11 
language-renpy 1.6.0 
scrolloff 0.2.0 
```

**Comment by Ben3eeE** [comment]
@itaniknight Would you mind opening a new issue in the [atom-keymap](https://github.com/atom/atom-keymap/issues) repository for the issue you are facing? Feel free to @mention me in the new issue.

Additional information that might be helpful in the new issue:

* As a workaround can you try adding a `Mode_switch` key for these four keys ASDF (QWERTY)? You currently have two of them.
* What do you expect to happen when you press `Mode_switch` and a key that you did not add in your `.Xmodmap`?

This issue was specific to Atom failing to build the keymap for certain languages and the issue you are having is different.

Thanks.

**Comment by Nrohtgnohp** [comment]
Still error on Ubuntu ( Atom version 1.13.1 ). I try to use Thai (TIS-620) not working.
Please fix or advice me for solution.

**Comment by 50Wliu** [comment]
@Nrohtgnohp as @as-cii mentioned above, this was fixed only in Atom 1.14.0.

**Comment by Nrohtgnohp** [comment]
But I cannot update to 1.14.0. I download from Atom main page can get only 1.13.1 version.

**Comment by 50Wliu** [comment]
https://atom.io/beta

**Comment by Nrohtgnohp** [comment]
Everything ok, Thanks.
